### PR TITLE
tun: add method for disabling TCP GRO on Linux

### PR DESCRIPTION
**Pull request by jwhited**
torvalds/linux@e269d79c7d35aa3808b1f3c1737d63dab504ddc8 broke virtio_net TCP & UDP GRO causing GRO writes to return EINVAL. The bug was then resolved later in
torvalds/linux@89add40066f9ed9abe5f7f886fe5789ff7e0c50e. The offending commit was pulled into various LTS releases.

Updates tailscale/tailscale#13041

### CODE REVIEW

File: tun/tun_linux.go [hunk 1]
```diff
@@ -464,7 +486,13 @@ func (tun *NativeTun) BatchSize() int {
 
 func (tun *NativeTun) DisableUDPGRO() {
 	tun.writeOpMu.Lock()
-	tun.udpGRO = false
+	tun.gro.disableUDPGRO()
+	tun.writeOpMu.Unlock()
+}
+
+func (tun *NativeTun) DisableTCPGRO() {
```
Comment (bradfitz) [inline tun/tun_linux.go]: Docs on these two? Guidance for when callers would want to call them?

(aside: is "GRO" even defined anywhere in the godoc, btw? That probably makes these methods look even more mysterious.)

File: tun/tun.go [hunk 2]
```diff
@@ -52,10 +52,14 @@ type Device interface {
 	BatchSize() int
 }
 
-type LinuxDevice interface {
+// GRODevice is a Device extended with methods for disabling GRO. Certain OS
+// versions may have offload bugs. Where these bugs negatively impact throughput
+// or break connectivity entirely we can use these methods to disable the
+// related offload.
```
Comment (bradfitz) [inline tun/tun.go:58]: wouldn't hurt to add a paragraph to this godoc with some more details (or a URL to more details) about the known problem(s) and their commit ranges/versions.

For upstreaming, a paragraph without a Tailscale URL is probably best?

**Review by bradfitz — APPROVED** [review]

File: tun/tun.go (hunk 2 above)
Comment (bradfitz) [reply tun/tun.go:58]: (The commit message is great, btw. Maybe just mirror that, but also include Linux kernel versions, if known? As LTS complicates it.)

File: tun/tun.go [hunk 3]
```diff
@@ -52,10 +52,14 @@ type Device interface {
...
+type GRODevice interface {
 	Device
-	// DisableUDPGRO disables UDP GRO if it is enabled. Certain device drivers
-	// (e.g. vxlan, geneve) do not properly handle coalesced UDP packets later
-	// in the stack, resulting in packet loss.
+	// DisableUDPGRO disables UDP GRO if it is enabled.
 	DisableUDPGRO()
+	// DisableTCPGRO disables TCP GRO if it is enabled.
+	DisableTCPGRO()
```
Comment (sailorfrag) [inline tun/tun.go:75]: Does anything currently call this?

Comment (sailorfrag) [reply tun/tun.go:75]: Nevermind, found it in the other PR

File: tun/tun.go (hunk 2 above)
Comment (jwhited) [reply tun/tun.go:58]: added in https://github.com/tailscale/wireguard-go/pull/33/commits/615c928261240e0007566e9f0285a589ea28698c

File: tun/tun_linux.go (hunk 1 above)
Comment (jwhited) [reply tun/tun_linux.go]: added in https://github.com/tailscale/wireguard-go/pull/33/commits/615c928261240e0007566e9f0285a589ea28698c

**Review by sailorfrag — APPROVED** [review]

**Pull request merged**